node/src/components/linear_chain_sync/operations.rs

```diff
@@ -263,16 +269,120 @@ where
     return Ok(outstanding_tries);
 }
 
+struct BlockWithMetadataFetchedFromRemote<I> {
+    block_with_metadata: Box<BlockWithMetadata>,
+    sender: I,
+}
+
+/// Fetches a block from the network by height.
+async fn fetch_and_store_block_by_height_from_remote<REv, I>(
+    effect_builder: EffectBuilder<REv>,
+    height: u64,
+    trusted_validator_weights: &BTreeMap<PublicKey, U512>,
+    finality_threshold_fraction: Ratio<u64>,
+) -> Result<Option<BlockWithMetadataFetchedFromRemote<I>>, FetcherError<BlockWithMetadata, I>>
+where
+    REv: From<FetcherRequest<I, BlockWithMetadata>>
+        + From<NetworkInfoRequest<I>>
+        + From<StorageRequest>,
+    I: Eq + Debug + Clone + Send + 'static,
+{
+    for peer in get_and_shuffle_network_peers(effect_builder).await {
+        match effect_builder
+            .fetch::<BlockWithMetadata, I>(height, peer.clone())
+            .await
+        {
+            Ok(FetchedData::FromStorage { item }) => {
+                // TODO: We do not know this peer has the block, don't use a fetcher
```

Comment:
Not sure I understand this comment. You mean we should basically set `sender: None` (i.e. make it an option)?

Reply:
Alternatively, we could just break out the code for fetching deploys, since fast sync does not need to actually validate blocks, just fetch deploys.